custom_components/pyscript/logbook.py

```diff
@@ -0,0 +1,45 @@
+"""Describe logbook events."""
+from homeassistant.core import callback
+
+from .const import DOMAIN
+
+import logging
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def async_describe_events(hass, async_describe_event):  # type: ignore
+    """Describe logbook events."""
+
+    @callback
+    def async_describe_logbook_event(event):  # type: ignore
+        """Describe a logbook event."""
+        data = event.data
+        func_args = data.get("func_args", {})
+        ev_name = data.get("name", "unknown")
+        ev_entity_id = data.get("entity_id", "pyscript.unknown")
+
+        ev_trigger_type = func_args.get("trigger_type", "unknown")
+        if ev_trigger_type == "event":
+            ev_source = f"event {func_args.get('event_type', 'unknown event')}"
+        elif ev_trigger_type == "state":
+            ev_source = f"state change {func_args.get('var_name', 'unknown entity')} == {func_args.get('value', 'unknown value')}"
+        elif ev_trigger_type == "time":
+            ev_trigger_time = func_args.get("trigger_time", "unknown")
+            if ev_trigger_time is None:
+                ev_trigger_time = "startup"
+            ev_source = f"time {ev_trigger_time}"
```

Comment:
Would it make sense (and be possible) to have a user-specified message that could override the defaults here?  That would allow the pyscript developer to optionally format their own messages that could be more informative based on the situation.

Related to that, is there a mechanism for pyscript to create its own logbook entries?

I'm happy to leave both of these issues for a future PR, rather than burden this one with supporting this too.  It is probably best to use the new features in this PR for a while before deciding how best to address user-defined logbook entries, or user-defined logbook messages for the builtin trigger, service and state logbook entries.

Reply:
To my knowledge, the only way to create a logbook entry is to call a service or fire an event. When you call a service, the logbook determines what that looks like on its own. When you fire an event, the "logbook.py" functionality here is the only thing I know of. Beyond that, the logbook.log service is available:

In pyscript...
```python
logbook.log(name="Kitchen", message="is being used", entity_id="domain.name", domain="light")
```

I will play with this and see if context is carried and the future service calls attributed to the logbook entry when done like this.

Allowing a user specified message would be awesome. If logbook.log allows that, then that's perfect. If not, I don't see any way to do it using the event.fire mechanism since we have to fire the event before the function even runs. Maybe with some decorator magic?

Reply:
In testing, `logbook.log` can generate any logbook message. However, if a subsequent service call is tied to it by context, then, it appears in the logbook as "test_2 turned on by service logbook.log" which is not useful at all.

So, users can use `logbook.log` to be more verbose about what their scripts are doing if desired, but the event firing with the correct context (and subsequent contexts) is the only way to tie them together.

Additionally, I tried firing a properly crafted event without the corresponding function in `logbook.py`. With this, the event does not show up in the logbook, and the subsequent state changes are attributed only to the service call.